pkg/backup/backupsink/sst_sink_key_writer_test.go

```diff
@@ -0,0 +1,568 @@
+						{key: "a", value: gtSSTSizeVal, timestamp: 10},
+						{key: "c", timestamp: 10},
+					}),
+				newMVCCKeySet("a", "d").
+					withKVs([]kvAndTS{{key: "b", timestamp: 10}}),
+				newMVCCKeySet("c", "f").
+					withKVs([]kvAndTS{{key: "e", timestamp: 10}}),
+			},
+			flushedSpans: []roachpb.Spans{
+				{{Key: s2k0("a"), EndKey: s2k0("c")}},
+				{{Key: s2k0("c"), EndKey: s2k0("d")}},
+				{{Key: s2k0("a"), EndKey: s2k0("d")}},
+			},
+			unflushedSpans: []roachpb.Spans{
+				{{Key: s2k0("c"), EndKey: s2k0("f")}},
+			},
+		},
+		// Two spans that are contiguous with each other should be merged into one span.
+		{
+			name: "extend-metadata",
+			exportKVs: []*mvccKVSet{
+				newMVCCKeySet("a", "c").withKVs([]kvAndTS{
+					{key: "a", timestamp: 10}, {key: "b", timestamp: 10},
+				}),
+				newMVCCKeySet("c", "e").withKVs([]kvAndTS{
+					{key: "c", timestamp: 10}, {key: "d", timestamp: 10},
+				}),
+			},
+			unflushedSpans: []roachpb.Spans{
+				{{Key: s2k0("a"), EndKey: s2k0("e")}},
+			},
+		},
+		// If a span overlaps its previous span, a flush must occur first. Should not occur in
+		// backup compactions.
+		{
+			name: "flush-from-overlapping-ooo-spans",
+			exportKVs: []*mvccKVSet{
+				newMVCCKeySet("a", "c").withKVs([]kvAndTS{
+					{key: "a", timestamp: 10}, {key: "b", timestamp: 10},
+				}),
+				newMVCCKeySet("b", "e").withKVs([]kvAndTS{
+					{key: "b", timestamp: 10}, {key: "c", timestamp: 10}, {key: "d", timestamp: 10},
+				}),
+			},
+			flushedSpans: []roachpb.Spans{
+				{{Key: s2k0("a"), EndKey: s2k0("c")}},
+			},
+			unflushedSpans: []roachpb.Spans{
+				{{Key: s2k0("b"), EndKey: s2k0("e")}},
+			},
+		},
+		// If a span precedes the previous span but does not overlap, a flush must
+		// occur first. Should not occur in backup compactions.
+		{
+			name: "flush-from-non-overlapping-ooo-spans",
+			exportKVs: []*mvccKVSet{
+				newMVCCKeySet("c", "e").withKVs([]kvAndTS{
+					{key: "c", timestamp: 10}, {key: "d", timestamp: 10},
+				}),
+				newMVCCKeySet("a", "c").withKVs([]kvAndTS{
+					{key: "a", timestamp: 10}, {key: "b", timestamp: 10},
+				}),
+			},
+			flushedSpans: []roachpb.Spans{
+				{{Key: s2k0("c"), EndKey: s2k0("e")}},
+			},
+			unflushedSpans: []roachpb.Spans{
+				{{Key: s2k0("a"), EndKey: s2k0("c")}},
+			},
+		},
+		// If a span is non-contiguous with the previous span, the BackupManifest_File must be
+		// flushed.
+		{
+			name: "flush-file-from-non-contiguous-in-order-spans",
+			exportKVs: []*mvccKVSet{
+				newMVCCKeySet("a", "c").withKVs([]kvAndTS{
+					{key: "a", timestamp: 10}, {key: "b", timestamp: 10},
+				}),
+				newMVCCKeySet("d", "g").withKVs([]kvAndTS{
+					{key: "d", timestamp: 10}, {key: "e", timestamp: 10},
+				}),
+			},
+			unflushedSpans: []roachpb.Spans{
+				{
+					{Key: s2k0("a"), EndKey: s2k0("c")},
+					{Key: s2k0("d"), EndKey: s2k0("g")},
+				},
+			},
+		},
+		// If a span is non-contiguous with the previous span and the previous span's last key
```

Comment:
the prev file should flush regardless of if the next span is contiguous, right? Maybe add another test case there too?